File: src/replay.ts
// ─────────────────────────────────────────────────────────────────────────────
// REPLAY / RECONSTRUCTION ENGINE (M3) — the observability read surface.
//
// Reconstructs and "replays" a past run PURELY from its journal: the ordered
// step sequence (inputs/decisions, effects, timing, outcome), reconstructed from
// the portable journal — NOT from live Restate/substrate state.
//
// Substrate-independence is structural: this module reads everything through a
// JournalSource (journal-source.ts). The same code runs against the live SQLite
// journal AND against a journal imported from a JSONL export with nothing else
// running. Proving replay over an imported export == replay over the live
// journal is the portability gate.
//
// Capabilities:
//   reconstruct(source, runId)        → full ReplayedRun (steps, effects, state)
//   stateAt(source, runId, n)         → time-travel: state + effects AS OF step N
//   divergencePoints(source, runId)   → where forks branched off this run
//   assertReplayMatches(a, b)         → replay-divergence assertion (== none)
//
// Read-only. No effects. No substrate. No mutation.
// ─────────────────────────────────────────────────────────────────────────────

import type { JournalSource } from "./journal-source.js";
import type { JournalEntry, RunMeta } from "./step-model.js";

/** A single reconstructed step in a replayed run. */
export interface ReplayStep {
  readonly seq: number;
  readonly stepName: string;
  readonly kind: string;
  /** The recorded output of the step (the "decision"/result). */
  readonly output: unknown;
  readonly sideEffect: boolean;
  /** True if this step was seeded by a fork (replayed, not natively executed). */
  readonly seeded: boolean;
  /** Source run if seeded; null for native steps. */
  readonly seededFrom: string | null;
  readonly idemKey: string;
  readonly recordedAt: string;
  /** Milliseconds elapsed from the previous step's recordedAt (null for seq 1). */
  readonly elapsedMsFromPrev: number | null;
  /** Side effects fired at this step (matched by stepName). */
  readonly effects: ReplayEffect[];
}

/** A side effect as reconstructed from the journal/export. */
export interface ReplayEffect {
  readonly id: number;
  readonly stepName: string;
  readonly idemKey: string;
  readonly firedAt: string;
  readonly payload: unknown;
}

/** Where another run forked off this run (a divergence point). */
export interface DivergencePoint {
  /** The seq on THIS run at which a fork branched. */
  readonly seq: number;
  /** The forked run id. */
  readonly forkRunId: string;
  readonly forkTrajectory: string;
}

/** A fully reconstructed run — everything a replay UI / CLI renders. */
export interface ReplayedRun {
  readonly runId: string;
  readonly trajectory: string;
  readonly meta: RunMeta | undefined;
  readonly steps: readonly ReplayStep[];
  /** Total wall-clock span of the run (ms), from first to last recordedAt. */
  readonly totalElapsedMs: number;
  /** The run's final outcome (output of the last step), or null if empty. */
  readonly outcome: unknown;
  /** All side effects fired by this run, in order. */
  readonly effects: readonly ReplayEffect[];
  /** Points where forks branched off this run. */
  readonly divergencePoints: readonly DivergencePoint[];
  /** Where this run reconstructed FROM (origin string of the JournalSource). */
  readonly reconstructedFrom: string;
}

function elapsedMs(prev: string, cur: string): number | null {
  const a = Date.parse(prev);
  const b = Date.parse(cur);
  if (Number.isNaN(a) || Number.isNaN(b)) return null;
  return b - a;
}

/** Effects fired at a given step, matched by step name. */
function effectsAtStep(
  allEffects: ReplayEffect[],
  stepName: string,
): ReplayEffect[] {
  return allEffects.filter((e) => e.stepName === stepName);
}

/**
 * Reconstruct a full run from a JournalSource. The returned object is the exact
 * data the replay UI/CLI renders: the step sequence (decisions + effects +
 * timing), the outcome, and where forks diverged.
 */
export function reconstruct(source: JournalSource, runId: string): ReplayedRun {
  const entries = source.trajectory(runId);
  const meta = source.runMeta(runId);
  const rawEffects = source.effectsFor(runId);
  const effects: ReplayEffect[] = rawEffects.map((r) => ({
    id: r.id,
    stepName: r.step_name,
    idemKey: r.idem_key,
    firedAt: r.fired_at,
    payload: safeParse(r.payload),
  }));

  const steps: ReplayStep[] = entries.map((e, i) => {
    const prev = i > 0 ? entries[i - 1]! : null;
    return {
      seq: e.seq,
      stepName: e.stepName,
      kind: e.kind,
      output: e.output,
      sideEffect: e.sideEffect,
      seeded: e.seededFrom !== null,
      seededFrom: e.seededFrom,
      idemKey: e.idemKey,
      recordedAt: e.recordedAt,
      elapsedMsFromPrev: prev ? elapsedMs(prev.recordedAt, e.recordedAt) : null,
      effects: effectsAtStep(effects, e.stepName),
    };
  });

  const totalElapsedMs =
    entries.length >= 2
      ? elapsedMs(entries[0]!.recordedAt, entries[entries.length - 1]!.recordedAt) ?? 0
      : 0;

  return {
    runId,
    trajectory: meta?.trajectory ?? "main",
    meta,
    steps,
    totalElapsedMs,
    outcome: entries.length > 0 ? entries[entries.length - 1]!.output : null,
    effects,
    divergencePoints: divergencePoints(source, runId),
    reconstructedFrom: source.origin,
  };
}

function safeParse(raw: string): unknown {
  try {
    return JSON.parse(raw) as unknown;
  } catch {
    return raw;
  }
}

/** Points where forks branched off `runId` (its direct children's fork seqs). */
export function divergencePoints(
  source: JournalSource,
  runId: string,
): DivergencePoint[] {
  return source
    .childRuns(runId)
    .filter((c) => c.forkedAtSeq !== null)
    .map((c) => ({
      seq: c.forkedAtSeq!,
      forkRunId: c.runId,
      forkTrajectory: c.trajectory,
    }))
    .sort((a, b) => a.seq - b.seq || (a.forkRunId < b.forkRunId ? -1 : 1));
}

// ─── Time-travel ──────────────────────────────────────────────────────────────

/** State + effects of a run AS OF step N (inclusive). The time-travel view. */
export interface StateAsOf {
  readonly runId: string;
  /** The step index travelled to (1-based, inclusive). */
  readonly n: number;
  /** The maximum seq available for this run. */
  readonly maxSeq: number;
  /** Steps with seq <= n (what had happened by step N). */
  readonly steps: readonly ReplayStep[];
  /** The output of step N (the "current state" at that point), or null. */
  readonly currentOutput: unknown;
  /** Side effects fired at or before step N. */
  readonly effects: readonly ReplayEffect[];
  /** Forks that had already diverged at or before step N. */
  readonly divergedSoFar: readonly DivergencePoint[];
  readonly reconstructedFrom: string;
}

/**
 * Time-travel: reconstruct the run state and effect set AS OF step N (inclusive).
 * Throws if N is out of range. Pure read over the journal — the same answer from
 * a live journal or an imported export.
 */
export function stateAt(
  source: JournalSource,
  runId: string,
  n: number,
): StateAsOf {
  const full = reconstruct(source, runId);
  const maxSeq = full.steps.length > 0 ? full.steps[full.steps.length - 1]!.seq : 0;
  if (!Number.isInteger(n) || n < 1 || n > maxSeq) {
    throw new Error(
      `time-travel target step ${n} out of range (run ${runId} has steps 1..${maxSeq})`,
    );
  }
  const steps = full.steps.filter((s) => s.seq <= n);
  const effects = full.effects.filter((e) => {
    // an effect belongs to a step at seq <= n if its step is within the prefix
    return steps.some((s) => s.stepName === e.stepName);
  });
  const currentStep = steps.find((s) => s.seq === n);
  return {
    runId,
    n,
    maxSeq,
    steps,
    currentOutput: currentStep?.output ?? null,
    effects,
    divergedSoFar: full.divergencePoints.filter((d) => d.seq <= n),
    reconstructedFrom: source.origin,
  };
}

// ─── Replay-divergence assertion (the correctness gate) ───────────────────────

/** The result of comparing two reconstructions of the "same" run. */
export interface ReplayDivergence {
  /** True if the two reconstructions are byte-identical in step seq+output+effects. */
  readonly identical: boolean;
  /** Human-readable list of differences (empty when identical). */
  readonly differences: string[];
  readonly stepsA: number;
  readonly stepsB: number;
}

/**
 * Assert that a reconstruction `b` (e.g. replayed from an imported export with NO
 * substrate) EXACTLY matches reconstruction `a` (e.g. the live journal of the
 * run that actually happened). "Replay-divergence = none" means identical step
 * sequence, identical per-step output, and identical effect set.
 *
 * This is the M3 correctness gate: prove the replay reproduces what actually
 * happened, and prove the offline-from-export replay reproduces the live one.
 */
export function assertReplayMatches(
  a: ReplayedRun,
  b: ReplayedRun,
): ReplayDivergence {
  const differences: string[] = [];

  if (a.steps.length !== b.steps.length) {
    differences.push(
      `step count differs: a=${a.steps.length} b=${b.steps.length}`,
    );
  }
  const n = Math.max(a.steps.length, b.steps.length);
  for (let i = 0; i < n; i++) {
    const sa = a.steps[i];
    const sb = b.steps[i];
    if (!sa || !sb) {
      differences.push(`step index ${i} present on only one side`);
      continue;
    }
    if (sa.seq !== sb.seq) differences.push(`seq mismatch at ${i}: ${sa.seq} vs ${sb.seq}`);
    if (sa.stepName !== sb.stepName)
      differences.push(`stepName mismatch at seq ${sa.seq}: ${sa.stepName} vs ${sb.stepName}`);
    if (sa.kind !== sb.kind)
      differences.push(`kind mismatch at seq ${sa.seq}: ${sa.kind} vs ${sb.kind}`);
    if (JSON.stringify(sa.output) !== JSON.stringify(sb.output))
      differences.push(
        `output mismatch at seq ${sa.seq}: ${JSON.stringify(sa.output)} vs ${JSON.stringify(sb.output)}`,
      );
    if (sa.idemKey !== sb.idemKey)
      differences.push(`idemKey mismatch at seq ${sa.seq}: ${sa.idemKey} vs ${sb.idemKey}`);
  }

  // Effect-set comparison (by idemKey + payload; ignore auto-increment id and
  // pid, which legitimately differ across stores — correctness is the logical
  // effect set, not the physical row id).
  const ea = a.effects
    .map((e) => `${e.idemKey}|${JSON.stringify(e.payload)}`)
    .sort();
  const eb = b.effects
    .map((e) => `${e.idemKey}|${JSON.stringify(e.payload)}`)
    .sort();
  if (JSON.stringify(ea) !== JSON.stringify(eb)) {
    differences.push(`effect set differs: a=${JSON.stringify(ea)} b=${JSON.stringify(eb)}`);
  }

  return {
    identical: differences.length === 0,
    differences,
    stepsA: a.steps.length,
    stepsB: b.steps.length,
  };
}

/** Convenience: reconstruct `runId` from both sources and assert they match. */
export function compareSources(
  liveSource: JournalSource,
  importedSource: JournalSource,
  runId: string,
): ReplayDivergence {
  return assertReplayMatches(
    reconstruct(liveSource, runId),
    reconstruct(importedSource, runId),
  );
}

/** Re-export for callers building the read surface. */
export type { JournalEntry };

File: python/durabl/replay.py
"""Read-only replay over a :class:`JournalSource` — mirrors ``src/replay.ts`` (subset)."""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime
from typing import Any

from durabl.journal_source import JournalSource
from durabl.types import EffectRow, JournalEntry, RunMeta


@dataclass(frozen=True)
class ReplayEffect:
    id: int
    step_name: str
    idem_key: str
    fired_at: str
    payload: Any


@dataclass(frozen=True)
class ReplayStep:
    seq: int
    step_name: str
    kind: str
    output: Any
    side_effect: bool
    seeded: bool
    seeded_from: str | None
    idem_key: str
    recorded_at: str
    elapsed_ms_from_prev: int | None
    effects: tuple[ReplayEffect, ...]


@dataclass(frozen=True)
class DivergencePoint:
    seq: int
    fork_run_id: str
    fork_trajectory: str


@dataclass(frozen=True)
class ReplayedRun:
    run_id: str
    trajectory: str
    meta: RunMeta | None
    steps: tuple[ReplayStep, ...]
    total_elapsed_ms: int
    outcome: Any
    effects: tuple[ReplayEffect, ...]
    divergence_points: tuple[DivergencePoint, ...]
    reconstructed_from: str


@dataclass(frozen=True)
class StateAsOf:
    run_id: str
    n: int
    max_seq: int
    steps: tuple[ReplayStep, ...]
    current_output: Any
    effects: tuple[ReplayEffect, ...]
    diverged_so_far: tuple[DivergencePoint, ...]
    reconstructed_from: str


def _elapsed_ms(prev: str, cur: str) -> int | None:
    try:
        a = datetime.fromisoformat(prev.replace("Z", "+00:00"))
        b = datetime.fromisoformat(cur.replace("Z", "+00:00"))
    except ValueError:
        return None
    return int((b - a).total_seconds() * 1000)


def _safe_parse(raw: str) -> Any:
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return raw


def _map_effects(rows: list[EffectRow]) -> list[ReplayEffect]:
    return [
        ReplayEffect(
            id=r.id,
            step_name=r.step_name,
            idem_key=r.idem_key,
            fired_at=r.fired_at,
            payload=_safe_parse(r.payload),
        )
        for r in rows
    ]


def _effects_at_step(effects: list[ReplayEffect], step_name: str) -> tuple[ReplayEffect, ...]:
    return tuple(e for e in effects if e.step_name == step_name)


def divergence_points(source: JournalSource, run_id: str) -> list[DivergencePoint]:
    out: list[DivergencePoint] = []
    for child in source.child_runs(run_id):
        if child.forked_at_seq is not None:
            out.append(
                DivergencePoint(
                    seq=child.forked_at_seq,
                    fork_run_id=child.run_id,
                    fork_trajectory=child.trajectory,
                )
            )
    out.sort(key=lambda d: (d.seq, d.fork_run_id))
    return out


def reconstruct(source: JournalSource, run_id: str) -> ReplayedRun:
    entries = source.trajectory(run_id)
    meta = source.run_meta(run_id)
    effects = _map_effects(source.effects_for(run_id))

    steps: list[ReplayStep] = []
    for i, e in enumerate(entries):
        prev: JournalEntry | None = entries[i - 1] if i > 0 else None
        elapsed = (
            _elapsed_ms(prev.recorded_at, e.recorded_at) if prev is not None else None
        )
        steps.append(
            ReplayStep(
                seq=e.seq,
                step_name=e.step_name,
                kind=e.kind,
                output=e.output,
                side_effect=e.side_effect,
                seeded=e.seeded_from is not None,
                seeded_from=e.seeded_from,
                idem_key=e.idem_key,
                recorded_at=e.recorded_at,
                elapsed_ms_from_prev=elapsed,
                effects=_effects_at_step(effects, e.step_name),
            )
        )

    total = 0
    if len(entries) >= 2:
        elapsed = _elapsed_ms(entries[0].recorded_at, entries[-1].recorded_at)
        total = elapsed if elapsed is not None else 0

    return ReplayedRun(
        run_id=run_id,
        trajectory=meta.trajectory if meta else "main",
        meta=meta,
        steps=tuple(steps),
        total_elapsed_ms=total,
        outcome=entries[-1].output if entries else None,
        effects=tuple(effects),
        divergence_points=tuple(divergence_points(source, run_id)),
        reconstructed_from=source.origin,
    )


def state_at(source: JournalSource, run_id: str, n: int) -> StateAsOf:
    full = reconstruct(source, run_id)
    max_seq = full.steps[-1].seq if full.steps else 0
    if not isinstance(n, int) or n < 1 or n > max_seq:
        raise ValueError(
            f"time-travel target step {n} out of range (run {run_id} has steps 1..{max_seq})"
        )
    steps = tuple(s for s in full.steps if s.seq <= n)
    effects = tuple(
        e for e in full.effects if any(s.step_name == e.step_name for s in steps)
    )
    current = next((s for s in steps if s.seq == n), None)
    return StateAsOf(
        run_id=run_id,
        n=n,
        max_seq=max_seq,
        steps=steps,
        current_output=current.output if current else None,
        effects=effects,
        diverged_so_far=tuple(d for d in full.divergence_points if d.seq <= n),
        reconstructed_from=source.origin,
    )
